### PR TITLE
traffic-gen: Get pci env var by name

### PR DESCRIPTION
**Pull request by RamLavi**
On [sriov-network-device-plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) a new label was [added](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/456) to the container env vars. 
This is causing the checkup to fail the traffic generator pod initiation.
The reason is the currently the trex config script is trying to find the PCI-address env
variable, assuming that there is only one env var that fits the description.
Since this assumption is wrong, changing the script to use the exact
name of the env variable.

### CODE REVIEW

**Review by orelmisan — CHANGES_REQUESTED** [review]
Thank you for the PR @RamLavi.
As we've discussed offline, this implementation will probably not solve [BZ 2177668](https://bugzilla.redhat.com/show_bug.cgi?id=2177668).

The SR-IOV device plugin uses the resourceName when it passes the PCI addresses of the VFs to kubelet as environment variables: https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#pod-device-information.

We need to use the NetworkAttachmentDefinition's resourceName:
```
apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      k8s.v1.cni.cncf.io/resourceName: openshift.io/intel_nics_dpdk
```

File: pkg/internal/checkup/checkup.go [hunk 1]
```diff
@@ -259,6 +265,32 @@ func (c *Checkup) createTrafficGeneratorPod(ctx context.Context) error {
...
+	}
+
+	const (
+		pciDevicesVarPrefix    = "PCIDEVICE"
+		pciDevicesVarDelimiter = "_"
+	)
+
+	// build the PCI Device env var naming recipe
+	// https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#pod-device-information
+	pciDevicesVarName := pciDevicesVarPrefix + pciDevicesVarDelimiter + rawResourceName
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup.go:288]: The variable name should also be upper-cased:
https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/blob/5f8442e5b92f15e5f63d6352efbf1b2e1108b574/pkg/resources/pool_stub.go#L118

File: pkg/internal/checkup/checkup_test.go [hunk 2]
```diff
@@ -335,6 +337,17 @@ func (cs *clientStub) GetPod(_ context.Context, namespace, name string) (*k8scor
...
 }
 
+func (cs *clientStub) GetNetworkAttachmentDefinition(_ context.Context, _, _ string) (*networkv1.NetworkAttachmentDefinition, error) {
+	return &networkv1.NetworkAttachmentDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"k8s.v1.cni.cncf.io/resourceName": "openshift.io/dpdk-net",
+			},
+		},
+		Spec: networkv1.NetworkAttachmentDefinitionSpec{},
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup_test.go:347]: This member is not a pointer, so it could be initialized implicitly.

**Review by orelmisan — COMMENTED** [review]
Thank you for the changes @RamLavi.
Could you please update the PR's description with a short explanation of the problem and the suggested solution?
Could you also please add a link to the BZ?

File: tests/checkup_test.go [hunk 3]
```diff
@@ -264,6 +264,11 @@ func newKubeVirtDPDKCheckerRole() *rbacv1.Role {
 				Resources: []string{"pods/exec"},
 				Verbs:     []string{"create"},
 			},
+			{
+				APIGroups: []string{"k8s.cni.cncf.io"},
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply tests/checkup_test.go:268]: DONE

File: pkg/internal/client/client.go [hunk 4]
```diff
@@ -24,18 +24,22 @@ import (
...
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/remotecommand"
 
 	kvcorev1 "kubevirt.io/api/core/v1"
+	netattdefclient "kubevirt.io/client-go/generated/network-attachment-definition-client/clientset/versioned/typed/k8s.cni.cncf.io/v1"
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/internal/client/client.go:36]: certainly. DONE

File: pkg/internal/client/client.go [hunk 5]
```diff
@@ -61,7 +65,12 @@ func New() (*Client, error) {
 		return nil, err
 	}
 
-	return &Client{client, config}, nil
+	cniClient, err := netattdefclient.NewForConfig(config)
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/internal/client/client.go:68]: not sure what you're asking.

File: README.md [hunk 6]
```diff
@@ -52,6 +52,9 @@ rules:
   - apiGroups: [ "" ]
     resources: [ "pods/exec" ]
     verbs: [ "create" ]
+  - apiGroups: [ "k8s.cni.cncf.io" ]
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply README.md:55]: np

File: pkg/internal/checkup/checkup.go (hunk 1 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:288]: yeah I saw it as soon as I ran it.. Fixed

File: pkg/internal/checkup/checkup_test.go (hunk 2 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup_test.go:347]: DONE

**Comment by RamLavi** [comment]
CI still fail (but for another reason). 
PR seems to do the job: 
```
+ cat /etc/trex_cfg.yaml
- port_limit: 2
  version: 2
  interfaces:
    - "0000:19:0a.0"
    - "0000:19:0a.1"
  port_bandwidth_gb: 10
  port_info:
    - ip: 10.10.10.2
      default_gw: 10.10.10.1
    - ip: 10.10.20.2
      default_gw: 10.10.20.1
  platform:
    master_thread_id: 2
    latency_thread_id: 42
    dual_if:
      - socket: 0
        threads: [4,6,8,44,46,48]
```

**Comment by RamLavi** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/8827f61a4e68da367928eab11d84f94a13875b98..aadaa6540034e80cbad258bce340e17c4ff6f800): Review fixes
@orelmisan

File: pkg/internal/client/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/internal/client/client.go:68]: Never mind.
It does not stop us from moving forward.

File: pkg/internal/checkup/checkup.go [hunk 7]
```diff
@@ -402,6 +435,10 @@ func newTrafficGeneratorPod(checkupConfig config.Config, secondaryNetworkRequest
 		dstWestMACAddressParamName: checkupConfig.DPDKWestMacAddress.String(),
 		dstEastMACAddressParamName: checkupConfig.DPDKEastMacAddress.String(),
 		verboseParamName:           strings.ToUpper(strconv.FormatBool(checkupConfig.Verbose)),
+		PCIDeviceVarParamName:      pciDevicesVarName,
+	}
+	if checkupConfig.Verbose {
+		log.Printf("envVars: %v", envVars)
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup.go:441]: Could you please split this change to a commit of its own?

Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:441]: DONE

**Comment by RamLavi** [comment]
passes on CNV4.12 cluster:
```
$ TEST_NAMESPACE=ralavi-ns make TEST_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest TRAFFIC_GEN_IMAGE_URL=quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest test/e2e 
docker run --rm \
           --volume /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:/root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:Z \
           --volume /root/.kube:/root/.kube:Z \
           --workdir /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup \
           -e KUBECONFIG=/root/.kube/config \
           -e TEST_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest \
           -e TEST_NAMESPACE=ralavi-ns \
           -e NETWORK_ATTACHMENT_DEFINITION_NAME= \
           -e RUNTIME_CLASS_NAME= \
           -e TRAFFIC_GEN_IMAGE_URL=quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest \
           -e VM_CONTAINER_DISK_IMAGE_URL= \
           docker.io/library/golang:1.19.4 go test ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h
=== RUN   TestKubevirtDpdkCheckup
Running Suite: KubevirtDpdkCheckup Suite - /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests
=======================================================================================================
Random Seed: 1678781330

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/test_suite_test.go:63
[BeforeSuite] PASSED [0.004 seconds]
------------------------------
Execute the checkup Job should complete successfully
/root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:79
  [FAILED] in [It] - /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89 @ 03/14/23 08:11:53.154
• [FAILED] [183.264 seconds]
Execute the checkup Job [It] should complete successfully
/root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:79

  [FAILED] checkup failed: {
  	"spec.param.networkAttachmentDefinitionName": "intel-dpdk-network-1",
  	"spec.param.testDuration": "1m",
  	"spec.param.trafficGeneratorImage": "quay.io/ramlavi/kubevirt-dpdk-checkup-traffic-gen:latest",
  	"spec.param.trafficGeneratorRuntimeClassName": "performance-performance-zeus10",
  	"spec.param.verbose": "true",
  	"spec.param.vmContainerDiskImage": "",
  	"spec.timeout": "10m",
  	"status.completionTimestamp": "2023-03-14T08:11:47Z",
  	"status.failureReason": "detected Error Packets on the traffic generator's side: Oerrors 0 Ierrors 5067",
  	"status.result.DPDKRxPacketDrops": "86317807",
  	"status.result.DPDKRxTestPackets": "573188359",
  	"status.result.DPDKTxPacketDrops": "0",
  	"status.result.DPDKVMNode": "zeus10.lab.eng.tlv2.redhat.com",
  	"status.result.trafficGeneratorInErrorPackets": "5067",
  	"status.result.trafficGeneratorNode": "zeus10.lab.eng.tlv2.redhat.com",
  	"status.result.trafficGeneratorOutputErrorPackets": "0",
  	"status.result.trafficGeneratorTxPackets": "840000006",
  	"status.startTimestamp": "2023-03-14T08:09:03Z",
  	"status.succeeded": "false"
  }
  In [It] at: /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89 @ 03/14/23 08:11:53.154
------------------------------

Summarizing 1 Failure:
  [FAIL] Execute the checkup Job [It] should complete successfully
  /root/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89

Ran 1 of 1 Specs in 183.269 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestKubevirtDpdkCheckup (183.27s)
FAIL
FAIL	github.com/kiagnose/kubevirt-dpdk-checkup/tests	183.301s
FAIL
make: *** [Makefile:59: test/e2e] Error 1
```

**Review by orelmisan — APPROVED** [review]
Thank you for the changes @RamLavi.

**Pull request merged**